pkg/services/ngalert/notifier/multiorg_alertmanager.go

```diff
@@ -78,49 +82,119 @@ func (moa *MultiOrgAlertmanager) LoadAndSyncAlertmanagersForOrgs(ctx context.Con
+	moa.alertmanagersMtx.Unlock()
+
+	// Now, we can stop the Alertmanagers without having to hold a lock.
+	for orgID, am := range amsToStop {
+		moa.logger.Info("stopping Alertmanager", "org", orgID)
+		am.StopAndWait()
+		moa.logger.Info("stopped Alertmanager", "org", orgID)
+	}
+}
+
+func (moa *MultiOrgAlertmanager) addOrUpdateAlertManagerConfigurations(orgIDs []int64, foundOrgs map[int64]struct{}, managerFactory alertManagerFactory) {
+	dbConfigs, err := moa.getLatestConfigs()
+	if err != nil {
+		moa.logger.Error("failed to load Alertmanager configurations", "err", err)
+		return
+	}
 	for _, orgID := range orgIDs {
-		orgsFound[orgID] = struct{}{}
-
-		existing, found := moa.alertmanagers[orgID]
-		if !found {
-			// These metrics are not exported by Grafana and are mostly a placeholder.
-			// To export them, we need to translate the metrics from each individual registry and,
-			// then aggregate them on the main registry.
-			m := metrics.NewAlertmanagerMetrics(moa.metrics.GetOrCreateOrgRegistry(orgID))
-			am, err := newAlertmanager(orgID, moa.settings, moa.configStore, moa.kvStore, m)
-			if err != nil {
-				moa.logger.Error("unable to create Alertmanager for org", "org", orgID, "err", err)
-			}
-			moa.alertmanagers[orgID] = am
-			existing = am
+		foundOrgs[orgID] = struct{}{}
+		dbConfig := dbConfigs[orgID]
+		moa.addOrUpdateAlertmanager(orgID, dbConfig, managerFactory)
+	}
+}
+
+// addOrUpdateAlertmanager gets an instance of AlertManager that serves a specific orgID from alertmanagers, and applies a new configuration.
+//If the configuration is not provided, it applies the default one.
+//If there is no AlertManager service the organization, it creates a new instance, applies the configuration (if provided, default otherwise) and adds it to alertmanagers.
+func (moa *MultiOrgAlertmanager) addOrUpdateAlertmanager(orgID int64, maybeDbConfig *models.AlertConfiguration, alertManagerFactory alertManagerFactory) {
+	var cfg *api.PostableUserConfig
+	if maybeDbConfig != nil {
+		var loadErr error
+		cfg, loadErr = Load([]byte(maybeDbConfig.AlertmanagerConfiguration))
+		if loadErr != nil {
+			moa.logger.Error("failed to parse Alertmanager config for org", "org", orgID, "id", maybeDbConfig.ID, "err", loadErr)
+			return
+		}
+	}
+
+	alertmanager, found := moa.alertmanagers[orgID]
+	if !found {
+		am, err := alertManagerFactory(orgID)
+		if err != nil {
+			moa.logger.Error("unable to create Alertmanager for org", "org", orgID, "err", err)
+			return
 		}
+		alertmanager = am
+	}
 
-		//TODO: This will create an N+1 query
-		if err := existing.SyncAndApplyConfigFromDatabase(); err != nil {
-			moa.logger.Error("failed to apply Alertmanager config for org", "org", orgID, "err", err)
+	if cfg != nil {
+		err := alertmanager.ApplyConfig(cfg)
+		if err != nil {
+			moa.logger.Error("failed to apply Alertmanager config for org", "org", orgID, "id", maybeDbConfig.ID, "err", err)
+			return
+		}
+	} else {
+		if found {
```

Comment:
Do we need an else condition here? What is the expected behaviour when the configuration is nil and organization could not be found? Is this an error?

Reply:
This covers a situation when the AM for an org already exists (the record in org table exists, and the instance was created during the previous iteration) but there is no configuration in database. 
I put warning to indicate of some unexpected but not critical condition.